### ISSUE TITLE
Add new type Blob to entity generator

### ISSUE THREAD

**Issue by andidev** [jhipster/generator-jhipster#1651]
Working on a PR

**Comment by phouverneyuff** [comment]
+1

**Comment by moifort** [comment]
:+1:

**Comment by gmarziou** [comment]
This is great because it can be hard to do it right with large objects.

There could be out of memory errors, issues about performance and streaming for reading and writing.

What about hibernate-validator, can it validate a blob max size or should a blob field be ignore in a validated entity?

**Comment by andidev** [comment]
@gmarziou 
Good point, If anyone knows how to handle size validation of blob fields please fill me in on this. I am quite unexperienced with blobs in databases so all suggestions are warmly welcome.

**Comment by andidev** [comment]
Any hints on what needs to be done in the mongodb version would also be appreciated. I guess all is needed is some serializer and deserializer annotation?

**Comment by jdubois** [comment]
@andidev not idea for MongoDB! And for Cassandra I should be more able to help, but I haven't done it yet with the new CQL Driver, only with Hector. -> in case of too much trouble, only make this option available for SQL, that's what most people use, so don't let the other databases slow you down.
As @gmarziou wrote, yes you need to validate the size, and yes that should be done in the controller. You can check [here](https://github.com/ippontech/tatami/blob/master/src/main/java/fr/ippon/tatami/web/fileupload/FileController.java#L256) how I did it a few years ago. That would be even better to do it with Bean Validation, but I don't know if that's possible.

**Comment by andidev** [comment]
So I have created an initial version but I'm not sure this solution is the correct way to handle blobs.
This is how it works:
The blob is included in the REST json when creating, updating the entity containing the blob. 
The blob is displayed by showing the base64 encoded content of the blob. Blob data is added by a html input tag with type="file". I had to create a directive to bind the file data to the ng-model.

No file name or type is saved when uploading a blob. This could be saved into the blob data as well or into another fields. Not sure what is the best way to do this and if it should be included. Any suggestions?

No size validation at the moment but I'll have a look into that later. Required validation is in place though.

Solution seems to work with mongodb as long as the content is less then 16MB (or something nearby)

Heres a gif demo of how to add and remove blob data (hope its not to large):
![blob-demo](https://cloud.githubusercontent.com/assets/571389/8398336/6eb0f330-1dea-11e5-9327-90ba62285454.gif)

@jdubois 
Could you review this solution before I go on creating the size validation?

**Comment by phouverneyuff** [comment]
Can I suggest this?
https://github.com/blueimp/jQuery-File-Upload

**Comment by gmarziou** [comment]
How does it compare to https://github.com/danialfarid/ng-file-upload ?

**Comment by phouverneyuff** [comment]
Good too. This is the demo:
https://blueimp.github.io/jQuery-File-Upload/angularjs.html

**Comment by deepu105** [comment]
There is ng flow as well which some one already added in a jhipster sample, i have it forked in my profile

**Comment by andidev** [comment]
I've managed implement the min and max byte validation on both angular side and java bean side.
I've also switched to using the ng-file-upload instead of a homemade one. This will make it more cross-browser compatible and support older browsers.

@deepu105 
ng-flow looks neat! Though it upload the file to the server directly after selecting it. My current implementation adds the blob data to the json and sends it with the rest data as any other field added to the generator entity. Not sure how to implement a simple general solution without adding any other fields than the byte array to the entity.
Any thoughts on how it could be implemented?

**Comment by gmarziou** [comment]
Is your implementation similar the [Spring MVC example](https://github.com/danialfarid/ng-file-upload/wiki/spring-mvc-example) in wiki?

**Comment by andidev** [comment]
No not really since it requires a HTTP multipart request. My implementation encodes the data in base64 format and sends the data in the json along with the REST request (the same way as any other field type generated with the entity generator).

This is how it currently looks in the html:
```html
<label translate="jhipsterApp.myEntety.aBlobField">ABlobField</label>
<div class="help-block clearfix">
    <span class="pull-left">{{abbreviate(myEntety.aBlobField)}}</span>
    <button ng-click="myEntety.aBlobField=null"
            ng-if="myEntety.aBlobField"
            class="btn btn-default btn-xs pull-right">
        <span class="glyphicon glyphicon-remove"></span>
    </button>
</div>
<input type="hidden" name="aBlobField" ng-model="myEntety.aBlobField" required minbytes="10" maxbytes="100">
<button type="file" ngf-select class="btn btn-default"
        ngf-change="setNgModel($files, myEntety)">Add File</button>

```

... and javascript:
```javascript
$scope.setNgModel = function ($files, myEntety) {
    console.log('$files = ', $files);
    if ($files[0]) {
        var file = $files[0];
        var fileReader = new FileReader();
        fileReader.readAsDataURL(file);
        fileReader.onload = function (e) {
            var data = e.target.result;
            var base64Data = data.substr(data.indexOf('base64,') + 'base64,'.length);
            $scope.$apply(function() {
                myEntety.aBlobField = base64Data;
            });
        };
    }
};
```

... and the entity:
```java
@Entity
@Table(name = "MYENTETY")
@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
public class MyEntety implements Serializable {

    ...

    @Lob
    @Size(min = 10, max = 100)
    @Column(name = "a_blob_field")
    private byte[] aBlobField;

    ...
}

```

the data is converted form a base64 encoded string into the entity `byte[]` field `aBlobField` by spring by default.

**Comment by gmarziou** [comment]
Mutlipart request encodes also in base64 so I guess it should be equivalent.

Have you tried large uploads?

As there's no chunking in client and no streaming on server, I suspect that you should hit a limit somewhere but I think it's OK because for very large files a database may not be the best place to store it anyway.

**Comment by andidev** [comment]
I have not tried large file uploads yet but I'll do that. However according to this [stackoverflow answer](http://stackoverflow.com/questions/1262376/is-there-a-limit-on-how-much-json-can-hold) there is no size limit in json requests, it's up to the server. Though I don't know what spring allows by default.

**Comment by gmarziou** [comment]
I guess that for very large objects, you could have network tileout or frop a reverse proxy if you use one and on server side you may have OutOfMemoryError depending on your JVM settings.

**Comment by andidev** [comment]
@gmarziou 
I have not tried experimenting with JVM memory sizes. Not sure where to set the limits and what the result will tell me. This blob field should not be used to store large files. The maximum limit I have found is around 250MB which is set by the javascript FileReader object for some reason.

@jdubois 
Could we merge this with a BETA tag?

**Comment by jdubois** [comment]
@andidev yes at the beginning we can just write BETA on the generator option, like we have for DTOs at the moment. At least people know it's not fully tested yet.
I'm a big fan of Linus' "release early, release often" mantra, and it's so easy to do a release with NPM!

**Comment by andidev** [comment]
Great! I'll add a BETA tag into the option in the generator right away.

It seems that elasticsearch indexes the blob content currently, anyone knows how to disable it? I guess it should be disabled.

**Comment by andidev** [comment]
link to PR https://github.com/jhipster/generator-jhipster/pull/1703

**Comment by jdubois** [comment]
Yes it should be disabled... I don't know how this is supposed to work, I expect there some kind of Spring Data ElasticSearch annotation?

**Comment by andidev** [comment]
Cant seem to find the correct annotation for doing that. Anyone that can help me out here?

**Comment by gzsombor** [comment]
Try @JsonIgnore, because the default serializer for Spring Data ElasticSearch use jackson under the hood.

**Comment by andidev** [comment]
But using Jsonignore will also exclude the blob from the json in REST calls, won't it?

**Comment by andidev** [comment]
Added generator option for content type (image or any)
Heres how the image crud api looks like
![blob-image-demo-mobile](https://cloud.githubusercontent.com/assets/571389/8613216/b4889926-26dc-11e5-8fcf-10b8497839c8.gif)

**Comment by gmarziou** [comment]
Great!

**Comment by gzsombor** [comment]
@andidev : yes, sure, this is one of the reason to use DTOs is to avoid these kind of problems.

**Comment by mourad91** [comment]
Good job @andidev :+1: 
@jdubois: it will be integrate on next version of jhipster ?

**Comment by andidev** [comment]
@mourad91
Yes it is merged with the master branch alredy, see PR #1703
